### PR TITLE
Ensure data chunks of a deleted composite blob get deleted

### LINKED ISSUE #298: Implement background/async deletes of defunct blobs
Defunct blobs are
1. successfully put chunks of PUT operations that eventually failed;
2. data chunks of deleted composite blobs (the deletes of which complete by just deleting the respective metadata chunks).

This will need some thinking through and a write-up before implementation.

### PR DESCRIPTION
**Pull request by pnarayanan**
Create a special OperationController that is responsible for
handling background delete operations in the router. Background
deletes are not latency sensitive and are best-effort. This
OperationController will use its own resources so as to not
interfere with regular operations.

Fixes #298 (approach described in [this](https://github.com/linkedin/ambry/issues/298#issuecomment-262604991) comment)

Coverage information (relevant newly added code is getting covered):

```
DeleteManager                   100% (2/ 2)     100% (11/ 11)   92.8% (77/ 83)
DeleteOperation                 100% (3/ 3)     100% (21/ 21)   95.4% (103/ 108)
FutureResult                    100% (1/ 1)     84.6% (11/ 13)  86.4% (19/ 22)
GetBlobInfoOperation            100% (2/ 2)     92.9% (13/ 14)  89.2% (99/ 111)
GetBlobOperation                100% (7/ 7)     98.1% (52/ 53)  94.2% (307/ 326)
GetBlobOptionsInternal          100% (1/ 1)     100% (1/ 1)     100% (3/ 3)
GetBlobResult                   100% (1/ 1)     100% (3/ 3)     100% (5/ 5)
GetBlobResultInternal           100% (1/ 1)     100% (1/ 1)     100% (3/ 3)
GetManager                      100% (2/ 2)     100% (12/ 12)   87% (67/ 77)
GetOperation                    100% (1/ 1)     100% (10/ 10)   100% (29/ 29)
GetRequestInfo                  100% (1/ 1)     100% (1/ 1)     100% (3/ 3)
NonBlockingRouter               100% (8/ 8)     97.4% (38/ 39)  83.2% (178/ 214)
NonBlockingRouterFactory        100% (1/ 1)     100% (3/ 3)     87% (20/ 23)
NonBlockingRouterMetrics        100% (7/ 7)     81% (17/ 21)    93.7% (149/ 159)
OperationCompleteCallback       100% (1/ 1)     100% (3/ 3)     100% (11/ 11)
```

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 1]
```diff
@@ -85,10 +89,13 @@
...
-    ocList = new ArrayList<OperationController>(routerConfig.routerScalingUnitCount);
-    for (int i = 0; i < routerConfig.routerScalingUnitCount; i++) {
-      ocList.add(new OperationController(i));
+    ocCount = routerConfig.routerScalingUnitCount;
+    ocList = new ArrayList<>();
+    for (int i = 0; i < ocCount; i++) {
+      ocList.add(new OperationController(Integer.toString(i)));
     }
+    backgroundDeleter = new BackgroundDeleter();
+    ocList.add(backgroundDeleter);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: if the purpose of adding to `ocList` is only for shutdown, I would recommend against it. Since there is only instance of `BackgroundDeleter`, it should be easy to shutdown/wait for shutdown of that one. I feel that keeping the _real_ ocs separated from the utility ocs is useful - one e.g. are the changes required for `getOperationController()`.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 2]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+  private class BackgroundDeleter extends OperationController {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Instantiate the BackgroundDeleter
+     * @throws IOException if the associated {@link OperationController} throws one.
+     */
+    BackgroundDeleter() throws IOException {
+      super("backgroundDeleter");
+      super.putManager.close();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: minor: do you need the `super` qualifier here?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 3]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+      routerMetrics.onPutBlobError(routerException);
+      operationCompleteCallback.completeOperation(futureResult, callback, null, routerException);
+    }
+
+    /**
+     * Initiate the deletes of the data chunks associated with this blobId, if this blob turns out to be a composite
+     * blob.
+     * @param blobId the blobId string associated with the possibly composite blob.
+     */
+    void initiateChunkDeletesIfAny(String blobId) {
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: so this essentially doubles the rate induced by delete at a server? Since Ambry's use case is such that deletes <<< gets, this should be ok, but could you call it out in the documentation of delete somewhere?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 4]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+      Callback<List<StoreKey>> callback = new Callback<List<StoreKey>>() {
+        @Override
+        public void onCompletion(List<StoreKey> result, Exception exception) {
+          if (exception != null) {
+            logger.error("Encountered exception when attempting to get chunks of a possibly composite deleted blob",
+                exception);
+          } else if (result != null) {
+            // result can be null if the blob is not a composite blob.
+            for (StoreKey key : result) {
+              currentOperationsCount.incrementAndGet(); // @todo
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: what's the todo here? Trying to keep background deletes out of the main count?

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java [hunk 5]
```diff
@@ -148,6 +149,11 @@ public void onCompletion(Void result, Exception exception) {
             for (Future future : futures) {
               future.get(AWAIT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
             }
+            long waitStart = SystemTime.getInstance().milliseconds();
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: why is this needed here? does the `DeleteManagerTest` test any composite blob deletes?

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 6]
```diff
@@ -395,6 +395,50 @@ public void onBlobDeleted(String blobId) {
...
+
+    setOperationParams();
+
+    String blobId = router.putBlob(putBlobProperties, putUserMetadata, putChannel).get();
+    router.deleteBlob(blobId, null).get();
+
+    long waitStart = SystemTime.getInstance().milliseconds();
+    while (router.getBackgroundOperationsCount() != 0
+        && SystemTime.getInstance().milliseconds() < waitStart + AWAIT_TIMEOUT_MS) {
+      Thread.sleep(1000);
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: not sure why you need this when you are doing `deletedDoneLatch.await()`. Could you explain?

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 7]
```diff
@@ -395,6 +395,50 @@ public void onBlobDeleted(String blobId) {
...
+    }
+
+    Assert.assertTrue("Deletes should not take longer than " + AWAIT_TIMEOUT_MS,
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: don't you also need to check that the right blob ids have been deleted?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 8]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+              currentBackgroundOperationsCount.incrementAndGet();
+              deleteManager.submitDeleteBlobOperation(key.getID(), new FutureResult<Void>(), new Callback<Void>() {
+                @Override
+                public void onCompletion(Void result, Exception exception) {
+                  if (exception != null) {
+                    logger.info("Background delete operation failed with exception", exception);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Don't you think this should be `warning` or `error`? `info` seems too benign for this.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 9]
```diff
@@ -395,6 +395,50 @@ public void onBlobDeleted(String blobId) {
   }
 
   /**
+   * Test that if a composite blob is deleted, the data chunks are eventually deleted.
+   */
+  @Test
+  public void testCompositeBlobDataChunksDelete() throws Exception {
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: if it is possible, could you also add test cases for deleting deleted blobs and deleting expired blobs?

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java [hunk 10]
```diff
@@ -252,3 +276,32 @@ private void removeAndAbort(GetOperation op, Exception abortCause) {
   }
 }
 
+/**
+ * Extension to {@link GetBlobOptions} containing additional parameters to the GetBlob operation that is internal to the
+ * router.
+ */
+class ExtendedGetBlobOptions extends GetBlobOptions {
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: `GetBlobOptions` is under ambry-api and is exposed to externals. We don't usually have private constructors and such in these classes and I am not very comfortable with that either. It is not the right place to include things internal to the router - and going forward the router may want to pass around more "options" internally. The other choice was to pass in yet another parameter to the methods taking the `GetBlobOptions`, but I felt that having an internal `Options` class would help in the future as well. 

Having a builder is orthogonal to all this.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 11]
```diff
@@ -46,6 +47,8 @@
 class NonBlockingRouter implements Router {
   private final NetworkClientFactory networkClientFactory;
   private final ArrayList<OperationController> ocList;
+  private final BackgroundDeleter backgroundDeleter;
+  private final int ocCount;
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Felt that it minimizes code bloat since this field of the `routerConfig` is used at a few places.

File: ambry-router/src/main/java/com.github.ambry.router/GetOperation.java [hunk 12]
```diff
@@ -41,10 +41,10 @@
   protected final NonBlockingRouterMetrics routerMetrics;
   protected final ClusterMap clusterMap;
   protected final ResponseHandler responseHandler;
-  protected final FutureResult<GetBlobResult> operationFuture;
-  protected final Callback<GetBlobResult> operationCallback;
+  protected final FutureResult operationFuture;
```
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: I guess the reason is there is no common ground between the two return types.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 4 above)
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Not related to this pr. I found `currentOperationsCount` is named a bit confusing. My first understanding is that it is a gauge-like metric that records the current operations, but it seems a so-far operation metric.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 13]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+      FutureResult<List<StoreKey>> futureResult = new FutureResult<>();
+      Callback<List<StoreKey>> callback = new Callback<List<StoreKey>>() {
+        @Override
+        public void onCompletion(List<StoreKey> result, Exception exception) {
+          if (exception != null) {
+            logger.error("Encountered exception when attempting to get chunks of a possibly composite deleted blob",
+                exception);
+          } else if (result != null) {
+            // result can be null if the blob is not a composite blob.
+            for (StoreKey key : result) {
```
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: In case of a large number of storekeys, is there a need to throttle the deletions?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 14]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+                }
+              });
+            }
+          }
+          currentBackgroundOperationsCount.decrementAndGet();
+        }
+      };
+      currentOperationsCount.incrementAndGet();
+      currentBackgroundOperationsCount.incrementAndGet();
+      getManager.submitGetChunkIdsOperation(blobId, futureResult, callback);
```
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Got a bit lost, I am wondering which part allows to get a blob after the blob is deleted?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 15]
```diff
@@ -482,6 +507,100 @@ public void run() {
...
+     * blob.
+     * @param blobId the blobId string associated with the possibly composite blob.
+     */
+    void initiateChunkDeletesIfAny(String blobId) {
+      FutureResult<List<StoreKey>> futureResult = new FutureResult<>();
+      Callback<List<StoreKey>> callback = new Callback<List<StoreKey>>() {
+        @Override
+        public void onCompletion(List<StoreKey> result, Exception exception) {
+          if (exception != null) {
+            logger.error("Encountered exception when attempting to get chunks of a possibly composite deleted blob",
```
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: log the blob id?

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java [hunk 16]
```diff
@@ -130,6 +132,28 @@ void submitGetBlobOperation(String blobId, GetBlobOptions options, FutureResult<
   }
 
   /**
+   * Submit an operation to get the chunk ids associated with this blob (if it is a composite blob), asynchronously.
+   * @param blobId The blobId for which the chunk ids are being requested.
+   * @param futureResult The {@link FutureResult} that contains the pending result of the operation.
+   * @param callback The {@link Callback} object to be called on completion of the operation.
+   */
+  void submitGetChunkIdsOperation(String blobId, FutureResult<List<StoreKey>> futureResult,
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: Yes - this could be merged into `submitGetBlobOperation`. Done.

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java (hunk 10 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: Discussed this offline with @vgkholla We concluded that this class is necessary. We also chose to make use of composition here. Made the change.

File: ambry-router/src/main/java/com.github.ambry.router/GetOperation.java (hunk 12 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: I have created a `GetBlobResultInternal` class for all types of results, so this brings back the typed future and callbacks. The instance itself will have the information about what kind of result it is.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 17]
```diff
@@ -25,6 +25,7 @@
 import com.github.ambry.notification.NotificationSystem;
 import com.github.ambry.protocol.RequestOrResponse;
 import com.github.ambry.protocol.RequestOrResponseType;
+import com.github.ambry.store.StoreKey;
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: I understand what you are saying. The thing is that the router can only get and fetch ids of type `BlobId`. Any change in ids have to go in as new versions of the `BlobId`. `StoreKey` does not embed any `partition` information and such. `StoreKey` is the abstraction at the server and the `MessageFormat` as they do not try to embed or get the partition information from the id.

Any inconsistency here is all due to the interaction with `MessageFormat`. Anything that is given or received to/from the `MessageFormat` has to be a `StoreKey`. The router also assumes that the chunk ids are `BlobId`s during a get - assuming that chunk ids are `BlobId`s if the metadata blob id is a `BlobId` is a fair assumption I think.

I will look into whether there is anything more we can do to make this usage cleaner and create an issue if needed.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 18]
```diff
@@ -294,6 +301,15 @@ int getOperationsCount() {
   }
 
   /**
+   * Return an approximate count of the number of background operations submitted to the router that are not yet
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: The reason is borrowed from `currentOperationsCount` which may not reflect an accurate number as certain operations fail before they actually get submitted and yet we increment this number (because the decrement happens in `completeOperation`). But I guess it is more confusing than helpful, so will remove the term - and we can say that incrementing this variable is when we consider an operation as active.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 19]
```diff
@@ -376,8 +393,16 @@ private void putBlob(BlobProperties blobProperties, byte[] userMetadata, Readabl
...
-    private void deleteBlob(String blobId, FutureResult<Void> futureResult, Callback<Void> callback) {
-      deleteManager.submitDeleteBlobOperation(blobId, futureResult, callback);
+    private void deleteBlob(final String blobId, FutureResult<Void> futureResult, final Callback<Void> callback) {
+      deleteManager.submitDeleteBlobOperation(blobId, futureResult, new Callback<Void>() {
+        @Override
+        public void onCompletion(Void result, Exception exception) {
+          if (exception == null) {
+            backgroundDeleter.initiateChunkDeletesIfAny(blobId);
+          }
+          callback.onCompletion(result, exception);
```
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: fixed.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: made `putManager` also a protected member, and removed the qualifier.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 3 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: will do.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 4 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: @xiahome Could you elaborate? This is not a metric. This is just an internal variable kept by the router - much like a size variable kept by a data structure, for example.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 8 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Sure, will log this as an `error`.

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java (hunk 5 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: No - but every delete will end up doing a background `get` now.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java (hunk 6 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: yeah, this is not needed.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 15 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Done.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 14 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: The `GetOptions` for this operation will use the `Include_All` flag.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java (hunk 7 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: Added.

**Comment by pnarayanan** [comment]
@vgkholla @xiahome Updated the PR to address your comments. Please take a look.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 4 above)
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: What I was trying to say is, if this counter is monotonically increasing, naming it `totalOperationsCount` seems more clear. Just a very minor comment.

Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: `currentOperationsCount` is not monotonically increasing - it is the number of active operations being processed by the router at any point in time.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 20]
```diff
@@ -395,6 +397,115 @@ public void onBlobDeleted(String blobId) {
...
+    router.deleteBlob(blobId, null).get();
+    Assert.assertTrue("Deletes should not take longer than " + AWAIT_TIMEOUT_MS,
+        deletesDoneLatch.get().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+
+    router.close();
+    assertClosed();
+    Assert.assertEquals("All operations should have completed", 0, router.getOperationsCount());
+  }
+
+  private Set<String> getBlobsInServers(MockServerLayout mockServerLayout) {
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: doc

**Comment by vgkholla** [comment]
@xiahome @pnarayanan said he will follow up with you if you have any comments

**Comment by xiahome** [comment]
LGTM. sorry a bit late.
